BUG: sparse: fix indexing after ellipsis and 2D array indexing

janezd opened the pull request:
#### Reference issue

Closes #21615.

#### What does this implement/fix?

When checking indices, `scipy.sparse._index.IndexMixin._validate_indices` skip ellipsis. This PR fixes it.

I apologize for not writing a test - I have problems building scipy on my machine.

janezd committed:
BUG: Fix incorrect indexing after ellipsis

Fixes: #21615.

If `a` is a sparse matrix, and we index it with `a[..., n]`,
`n` is compared with `a.shape[0]` instead of with `a.shape[1]`.

dschult committed:
add more tests for ellipsis

dschult committed:
add tests and fix 2d-index bug

add tests of 2d-index array in fancy indexing
fix bugs by replacing squeeze with ravel.
fix 2d index for spmatrix

Update test_base.py::test_inplace_ops to use self.asdense instead of .squeeze()

replace squeeze with ravel when indexing, fix 2d index for spmatrix, add tests

dschult commented:
I’ve added new additional tests for using ellipsis with integers, and slices. And we had none with ellipsis with arrays, so I added those too. That helped me realize that we didn’t have tests for array indexing where the array is a single value. Adding those caught two more bugs in the DOK and LIL formats. We also needed more tests of 2D arrays with indexing, which revealed spmatrix improper treatment.  These are each rarely used features, but together they plug some bugs and holes in testing.  The OP is a commonly used case, and ellipsis is a newly added feature. So thanks again for reporting and fixing it. I pushed the tests to your branch, so you’ll need to pull it to local before doing anything more (which you may not need to).

dschult reviewed (approved):
Summary:
- fixed out of range integer indexing alongside ellipsis.
- added many indexing tests with ellipsis. New tests cover index value (out of range issues)
- Added index ellipsis tests for array indexing alongside ellipsis (fancy indexing).
- Added tests of array indexing where the array has a single value e.g. A[[1], 2]
- Fixed bugs in format DOK and LIL spmatrix and using array indexing with single value.
- Adjusted `test_inplace_ops to use ` self.asdense`instead of `squeeze`.  

I approve this PR.    Thanks @janezd !

janezd committed:
add one more test for ellipsis

janezd commented:
@dschult, thanks. I apologize again for not contributing any tests - only today I finally managed to build scipy from sources.

I now added one test that is closer to the initial problem. Test that `a[..., 10]` fails didn't cover the case when the second index is within the range of the first dimension (against which it would be matched before this fix), but not within the second dimension. So I added a test that `a.T[..., 5]` fails, too. Just a detail - feel free to squash with other commits.

dschult commented:
Good point -- the example in the issue is about "index after `...` fits in the second but not in the first" (a 2x3 array with index 2 in 2nd position). But you are right that we could also check "index after `...` fits in the first but not in the second".

That requires changing the array sizes and transpose is a quick easy way to do that. Perhaps we could do that with all these cases, but different cases raise for different out-of-bounds. It'd be nice to have a function that compared what happens with sparse and with numpy (both raise or results are equal) that we could apply to the many cases tested here.

janezd commented:
A function would be nice, but there is no convenient syntax to pass a slice as an argument; I find `np.s_` ... noisy. I also don't like asserts within loops (like in `test_slicing_3`, for instance) because they are annoying to debug.

Maybe we should have a test object that can be indexed? Like

```python
class _TestSlicing:
    def create_tester(self, m):
        class EqTester:
            def __init__(self, self_, m):
                self.dense = m
                self.sparse = self_.spcreator(m)

            def __getitem__(self, idx):
                try:
                    dense_val = self.dense[idx]
                except Exception as exc:
                    with assert_raises(type(exc)):
                        self.sparse[idx]
                else:
                    assert_equal(dense_val, self.sparse[idx].toarray())  
        
        return EqTester(self, m)
```

This allows for writing tests like this:

```python
    def test_get_slices(self):
        assert_equal_result = self.create_tester(arange(50.).reshape(5,10))

        assert_equal_result[2:5, 0:3]
        assert_equal_result[1:, :-1]
        assert_equal_result[:-1, 1:]
        assert_equal_result[1000]  # raises the same exception
```

It may be confusing that tests are done through (a side effect of) indexing not with calls, but I think the name explains it - and the syntax is short and clean.

A drawback is that this does not check exception messages because they are different for dense and sparse matrices. This could be solved by passing the message as an extra "index" (not nice) or, perhaps better, by changing the messages for sparse matrices to match those for the dense.

I hope I'm not writing nonsense. My knowledge of scipy codebase is very limited, and I'm used to unittests, not pytest.

dschult commented:
I like it!  I don't think it even needs the outside class _TestSlicing because it stands alone.
I would also rename the function from `create_tester` to something like `compare_indexing_for` so the testing function could be:
```python
    def test_get_slices(self):
        a = arange(50.).reshape(5,10)
        for A in (a, a.T):
            assert_index_like_np = compare_indexing_for(A)

            assert_index_like_np[2:5, 0:3]
            assert_index_like_np[1:, :-1]
            assert_index_like_np[:-1, 1:]
            assert_index_like_np[1000]  # raises the same exception
```

janezd commented:
Your names are much better.

The outside class is the existing class `_TestSlicing` (or another one; we can also put this into its own mixin if we need to reuse it in different places). It needs some outer class to access `spcreator`.

If you think we'd often run the same tests for transposed matrices, we can avoid loops by building them into the test.

```python
class _TestSlicing:
    def compare_indexing_for(self, dense, transp=False):
        sparse_ = self.spcreator(dense)
        sparse_t = self.spcreator(dense.T) if transp else None

        def test_indices(d, s, idx):
            try:
                exp = d[idx]
            except Exception as exc:
                with assert_raises(type(exc)):
                    s[idx]
            else:
                assert_equal(exp, s[idx].toarray())  

        class Compare:
            def __getitem__(self, idx):
                test_indices(dense, sparse_, idx)
                if transp:
                    test_indices(dense.T, sparse_t, idx)

        return Compare()
```

So the test becomes

```python
    def test_get_slices(self):
        assert_index_like_np = self.compare_indexing_for(
            arange(50.).reshape(5,10),
            transp=True)

        assert_index_like_np[2:5, 0:3]
        assert_index_like_np[1:, :-1]
        assert_index_like_np[:-1, 1:]
        assert_index_like_np[1000]
```

What about error messages? Is there a reason that messages for dense and sparse matrices are different (`IndexError: index 1000 is out of bounds for axis 0 with size 5` vs `IndexError: row index (11) out of range`)? Would unifying them be a significant backward incompatible change?

If we want to keep them as they are but still test them, we can incorporate them into the test, like this,

```
        assert_index_like_np[1000, 'row index (11) out of range']
```

I'm not sure I like it much, plus it doesn't work for transposed matrices.

dschult commented:
I don't envision that testing specifically the transpose array will be used often (based on the tests currently in `test_base.py`). But there are several places where we test various indexing patterns on the same array. So this could be helpful.

I think changing the style of lots of tests is not the focus of this PR. So, maybe for this one we can construct the function and apply it in one test method.  Then after review and conversation we can decide how best to apply that elsewhere.

And it should work fine for it to be a function so long as we pass in `self`. Something like:
`compare_indexing_for(self, A)`
The function can get `spcreator` from that first argument.

The question about error messages is a longstanding "wart" of the library. In some important cases, we want the messages to differ for sparse than for numpy. But in many others they probably ought to be the same. And while it might be tempting to go and rewrite all the error messages some day, that will cause headaches for people who are relying on the error message text to run their own tests. So, I think we shouldn't try to fix that here in this PR. It's a bigger issue.

janezd commented:
> So, maybe for this one we can construct the function and apply it in one test method.

What about just keeping this PR as it is now? Its tests are written in the same style as the existing tests. If we have ideas how to rewrite (simplify, generalize) tests, let us not introduce functions that we might remove in a future PR anyway.

> And while it might be tempting to go and rewrite all the error messages some day, that will cause headaches for people who are relying on the error message text to run their own tests.

Sure, and I may well be one of those people with a headache. :) The whole reason I started this was that tests of our software on nightly built scipy wheels were failing.

dschult commented:
I agree -- let's merge this PR as is, and we can adjust the tests to include new tooling in another PR.
We've fixed 3 bugs, and added many new tests.

Thanks!

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -83,6 +83,8 @@ def __getitem__(self, key):
         else:
             if isinstance(col, INT_TYPES):
                 res = self._get_arrayXint(row, col)
+                if row.ndim > 1:
```
Let's move this logic into the `_get_arrayXint` method if possible.

dschult committed:
move index reshape inside _get_arrayXint
